### PR TITLE
Allow setting userId to nil

### PR DESCRIPTION
**Pull request by djih**
Added tests for initialization and clearing userId

### CODE REVIEW

File: Amplitude/Amplitude.m [hunk 1]
```diff
@@ -965,16 +971,15 @@ - (void)setUserProperties:(NSDictionary*) userProperties replace:(BOOL) replace
...
-    if (![self isArgument:userId validType:[NSString class] methodName:@"setUserId:"]) {
+    if (!([self isArgument:userId validType:[NSString class] methodName:@"setUserId:"] || userId == nil)) {
         return;
     }
     
     [self runOnBackgroundQueue:^{
-        (void) SAFE_ARC_RETAIN(userId);
         SAFE_ARC_RELEASE(_userId);
-        _userId = userId;
+        _userId = SAFE_ARC_RETAIN(userId);
```
Comment (curtisliu) [inline Amplitude/Amplitude.m]: I think the retain has to come before the release, otherwise userId could be freed before this is set. could be wrong though

File: Amplitude/Amplitude.m [hunk 2]
```diff
@@ -310,17 +320,13 @@ - (void)initializeApiKey:(NSString*) apiKey userId:(NSString*) userId
...
     _apiKey = apiKey;
     
     [self runOnBackgroundQueue:^{
         @synchronized (_eventsData) {
-            if (userId != nil) {
-                [self setUserId:userId];
-            } else {
-                _userId = SAFE_ARC_RETAIN([_eventsData objectForKey:@"user_id"]);
-            }
+            [self setUserId: setUserId ? userId : [_eventsData objectForKey:@"user_id"]];
```
Comment (curtisliu) [inline Amplitude/Amplitude.m]: i think this should only get called if setUserId is YES, otherwise it's a noop call that forces an unnecessary saveEventsData call

File: Amplitude/Amplitude.m (hunk 1 above)
Comment (djih) [inline Amplitude/Amplitude.m]: Good point. That's probably why it was the way it was. I'll change it back

File: Amplitude/Amplitude.m (hunk 2 above)
Comment (djih) [inline Amplitude/Amplitude.m]: Another good point. Change it back

**Comment by djih** [comment]
PTAL, okay to update cocopods version?

File: Amplitude/Amplitude.m [hunk 3]
```diff
@@ -310,13 +319,13 @@ - (void)initializeApiKey:(NSString*) apiKey userId:(NSString*) userId
         return;
     }
 
-    (void) SAFE_ARC_RETAIN(apiKey);
+    SAFE_ARC_RETAIN(apiKey);
```
Comment (curtisliu) [inline Amplitude/Amplitude.m]: does removing the (void) affect anything? It's an unrelated change that could be important

**Comment by curtisliu** [comment]
change the cocoapods version in another commit, not this PR

File: Amplitude/Amplitude.m (hunk 3 above)
Comment (djih) [inline Amplitude/Amplitude.m]: I don't think the (void) does anything. Retain returns self, but it's not being saved to a variable. But I'm not entirely sure if it won't affect anything, so I will add the (void) back in

**Comment by djih** [comment]
reverted void, cocopods version, PTAL

**Comment by curtisliu** [comment]
LGTM
